File: client/src/main/java/au/com/codeka/warworlds/client/game/welcome/SignInLayout.java
package au.com.codeka.warworlds.client.game.welcome;

import android.content.Context;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.TextView;

import au.com.codeka.warworlds.client.R;
import au.com.codeka.warworlds.client.util.GameSettings;
import au.com.codeka.warworlds.client.util.ViewBackgroundGenerator;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Layout for {@link SignInScreen}.
 */
public class SignInLayout extends RelativeLayout {
  interface Callbacks {
    void onSignInClick(String emailAddr);
    void onCancelClick();
  }

  private final Button signInButton;
  private final Button cancelButton;
  private final EditText emailText;
  private final TextView signInHelp;
  private final TextView signInError;

  public SignInLayout(Context context, Callbacks callbacks) {
    super(context);
    inflate(context, R.layout.signin, this);
    ViewBackgroundGenerator.setBackground(this);

    signInHelp = checkNotNull(findViewById(R.id.signin_help));
    signInError = checkNotNull(findViewById(R.id.signin_error));
    signInButton = checkNotNull(findViewById(R.id.signin_btn));
    cancelButton = checkNotNull(findViewById(R.id.cancel_btn));
    emailText = checkNotNull(findViewById(R.id.email));
    if (GameSettings.i.getString(GameSettings.Key.EMAIL_ADDR).isEmpty()) {
      signInButton.setText(R.string.next);
    } else {
      signInButton.setText(R.string.switch_user);
    }
    signInButton.setOnClickListener(v -> callbacks.onSignInClick(emailText.getText().toString()));

    cancelButton.setOnClickListener(v -> callbacks.onCancelClick());
  }

  /**
   * Update the state of the UI.
   *
   * @param helpResId The resource ID to use for the help text.
   * @param signInResId
   * @param isCancel
   * @param showEmailAddr
   */
  public void updateState(
      int helpResId,
      boolean signInEnabled,
      int signInResId,
      boolean isCancel,
      boolean showEmailAddr) {
    if (showEmailAddr) {
      emailText.requestFocus();
      emailText.setVisibility(View.VISIBLE);
    } else {
      emailText.setVisibility(View.GONE);
    }
    signInHelp.setText(helpResId);
    InputMethodManager imm =
        (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
    imm.showSoftInput(emailText, InputMethodManager.SHOW_IMPLICIT);
    signInButton.setEnabled(signInEnabled);
    signInButton.setText(signInResId);
    cancelButton.setText(isCancel ? R.string.cancel : R.string.back);
  }

  public void hideKeyboard() {
    InputMethodManager imm =
        (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
    if (imm != null) {
      imm.hideSoftInputFromWindow(emailText.getWindowToken(), 0);
    }
  }

  public void showKeyboard() {

  }

  public void setForceSignIn(int helpResId) {
    signInHelp.setText(helpResId);
    signInButton.setText(R.string.yes);
    cancelButton.setText(R.string.no);
  }

  public void setErrorMsg(int resId) {
    signInError.setText(resId);
  }
  public void setErrorMsg(String errorMsg) {
    if (errorMsg == null) {
      signInError.setText("");
    } else {
      signInError.setText(errorMsg);
    }
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/game/welcome/SignInLayout.kt
package au.com.codeka.warworlds.client.game.welcome

import android.content.Context
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.RelativeLayout
import android.widget.TextView
import au.com.codeka.warworlds.client.R
import au.com.codeka.warworlds.client.util.GameSettings
import au.com.codeka.warworlds.client.util.GameSettings.getString
import au.com.codeka.warworlds.client.util.ViewBackgroundGenerator.setBackground
import com.google.common.base.Preconditions

/**
 * Layout for [SignInScreen].
 */
class SignInLayout(context: Context?, callbacks: Callbacks) : RelativeLayout(context) {
  interface Callbacks {
    fun onSignInClick()
    fun onCreateEmpireClick()
    fun onCancelClick()
  }

  private val signInButton: Button
  private val cancelButton: Button
  private val signInHelp: TextView
  private val signInError: TextView

  private var inCreateEmpireError: Boolean = false

  init {
    View.inflate(context, R.layout.signin, this)
    setBackground(this)
    signInHelp = findViewById(R.id.signin_help)
    signInError = findViewById(R.id.signin_error)
    signInButton = findViewById(R.id.signin_btn)
    cancelButton = findViewById(R.id.cancel_btn)

    signInButton.setOnClickListener {
      if (inCreateEmpireError) {
        callbacks.onCreateEmpireClick()
      } else {
        callbacks.onSignInClick()
      }
    }
    cancelButton.setOnClickListener { callbacks.onCancelClick() }
  }

  /**
   * Update the state of the UI.
   *
   * @param helpResId The resource ID to use for the help text.
   * @param signInResId
   * @param isCancel
   */
  fun updateState(
      helpResId: Int,
      signInEnabled: Boolean,
      signInResId: Int,
      isCancel: Boolean) {
    signInHelp.setText(helpResId)

    signInButton.isEnabled = signInEnabled
    signInButton.setText(signInResId)
    cancelButton.setText(if (isCancel) R.string.cancel else R.string.back)
  }

  fun setForceSignIn(helpResId: Int) {
    signInHelp.setText(helpResId)
    signInButton.setText(R.string.yes)
    cancelButton.setText(R.string.no)
  }

  fun setCreateEmpireError(helpResId: Int, emailAddr: String?) {
    signInHelp.text = resources.getString(helpResId, emailAddr)
    signInButton.isEnabled = true
    signInButton.setText(R.string.create_empire)
    cancelButton.setText(R.string.cancel)
    inCreateEmpireError = true
  }

  fun setErrorMsg(resId: Int) {
    signInError.setText(resId)
  }

  fun setErrorMsg(errorMsg: String?) {
    if (errorMsg == null) {
      signInError.text = ""
    } else {
      signInError.text = errorMsg
    }
  }
}